google/cloud/bigtable/data/_metrics/handlers/opentelemetry.py

```diff
@@ -0,0 +1,196 @@
+        self.throttling_latencies = meter.create_histogram(
+            name="throttling_latencies",
+            description="The latency introduced by the client by blocking on sending more requests to the server when there are too many pending requests in bulk operations.",
+            unit="ms",
+        )
+
+
+class OpenTelemetryMetricsHandler(MetricsHandler):
+    """
+    Maintains a set of OpenTelemetry metrics for the Bigtable client library,
+    and updates them with each completed operation and attempt.
+
+    The OpenTelemetry metrics that are tracked are as follows:
+      - operation_latencies: latency of each client method call, over all of it's attempts.
+      - first_response_latencies: latency of receiving the first row in a ReadRows operation.
+      - attempt_latencies: latency of each client attempt RPC.
+      - retry_count: Number of additional RPCs sent after the initial attempt.
+      - server_latencies: latency recorded on the server side for each attempt.
+      - connectivity_error_count: number of attempts that failed to reach Google's network.
+      - application_latencies: the time spent waiting for the application to process the next response.
+      - throttling_latencies: latency introduced by waiting when there are too many outstanding requests in a bulk operation.
+    """
+
+    def __init__(
+        self,
+        *,
+        project_id: str,
+        instance_id: str,
+        table_id: str,
+        app_profile_id: str | None = None,
+        client_uid: str | None = None,
+        instruments: _OpenTelemetryInstruments = _OpenTelemetryInstruments(),
+        **kwargs,
+    ):
+        super().__init__()
+        self.otel = instruments
+        # fixed labels sent with each metric update
+        self.shared_labels = {
+            "client_name": f"python-bigtable/{bigtable_version}",
+            "client_uid": client_uid or str(uuid4()),
+            "resource_project": project_id,
+            "resource_instance": instance_id,
+            "resource_table": table_id,
+        }
+        if app_profile_id:
+            self.shared_labels["app_profile"] = app_profile_id
+
+    def on_operation_complete(self, op: CompletedOperationMetric) -> None:
+        """
+        Update the metrics associated with a completed operation:
+          - operation_latencies
+          - retry_count
+        """
+        try:
+            status = str(op.final_status.value[0])
+        except (IndexError, TypeError):
+            status = "2"  # unknown
+        labels = {
+            "method": op.op_type.value,
+            "status": status,
+            "resource_zone": op.zone,
+            "resource_cluster": op.cluster_id,
+            **self.shared_labels,
+        }
+        is_streaming = str(op.is_streaming)
+
+        self.otel.operation_latencies.record(
+            op.duration, {"streaming": is_streaming, **labels}
+        )
+        self.otel.retry_count.add(len(op.completed_attempts) - 1, labels)
```

Comment:
We leave retry_count as empty if there's no retries. So we only export this if (len(op.completed_attempt) - 1 > 0)

Reply:
Ok, so don't even send anything for this metric if there's no errors? And I assume that applies to connectivity_error_count too then?